src/sender_queue/mod.rs

```diff
@@ -248,10 +293,64 @@ where
         }
     }
 
+    /// Removes a given old participant if it has been scheduled for removal as a result of being
+    /// superseded by a new set of participants of which it is not a member. Returns `true` if the
+    /// participant has been removed and `false` otherwise.
+    fn remove_participant_if_old(&mut self, id: &D::NodeId) -> bool {
+        self.last_epochs
+            .get(id)
+            .cloned()
+            .map_or(false, |last_epoch| self.remove_participant(id, &last_epoch))
+    }
+
+    /// Removes a given old participant after a specified epoch if that participant has become
+    /// superseded by a new set of participants of which it is not a member. Returns `true` if the
+    /// participant has been removed and `false` otherwise.
+    fn remove_participant_after(&mut self, id: &D::NodeId, last_epoch: &D::Epoch) -> bool {
+        self.last_epochs.insert(id.clone(), last_epoch.clone());
+        self.remove_participant(id, last_epoch)
+    }
+
+    /// Removes a participant after a specified last epoch. The participant is removed if
+    ///
+    /// 1. its epoch is newer than its last epoch, or
+    ///
+    /// 2. the epoch of the managed algorithm instance is newer than the last epoch and the sender
+    /// queue has sent all messages for all epochs up to the last epoch to the participant.
+    ///
+    /// Returns `true` if the participant has been removed and `false` otherwise.
+    fn remove_participant(&mut self, id: &D::NodeId, last_epoch: &D::Epoch) -> bool {
+        let remove = if let Some(peer_epoch) = self.peer_epochs.get(id) {
+            last_epoch < peer_epoch
+                || (*last_epoch < self.algo.epoch() && self.outgoing_queue.get(id).map_or(
+                    false,
+                    |q| {
+                        q.keys()
+                            .take_while(|epoch| *epoch <= last_epoch)
+                            .next()
+                            .is_none()
+                    },
+                ))
+        } else {
+            true
+        };
```

Comment:
That's reeeally hard to read, but I also can't come up with anything much better right now.
It _might_ help (_or_ make it worse) to `return false` in all the places where we'd assign `false` to `remove`, and get rid of that variable. The last line of the method would then just be `true`. Not sure.
But I'm also fine with merging it like that for now. Maybe we can reformulate it later.